src/sources/mod.rs

```diff
@@ -85,10 +85,17 @@ impl BitOrAssign for PostAction {
 ///
 /// In case your event source needs to do some special processing before or after a
 /// polling session occurs (to prepare the underlying source for polling, and cleanup
-/// after that), you can override the `pre_run` and `post_run`, that do nothing by
-/// default. Depending on the underlying events, `process_events` may be invoked once,
-/// several times, or none at all between `pre_run` and `post_run` are called, but when
-/// it is invoked, it'll always be between those two.
+/// after that), you can override [`NEEDS_EXTRA_LIFECYCLE_EVENTS`] to `true`.
+/// For all sources for which that constant is `true`, the methods [`before_will_sleep`] and
+/// [`before_handle_events`] will be called.
+/// [`before_will_sleep`] is called before the polling system performs a poll operation.
+/// [`before_handle_events`] is called before any process_events methods have been called.
+/// This means that during `process_events` you can assume that all cleanup has occured on
+/// all sources.
```

Comment:
I'd prefer if this comment also gave some notice about how to handle this when building composite sources, so set the constant to `true` if at least one of the sub-sources need it, and forward the calls to the two methods to the sub-sources that need it.

Also, writing this, I realize that `before_will_sleep` would need to return a `Vec<(Readiness, Token))>` (or something similar), rather than just an option, to cover the case when a composite source has more than one sub-source that needs this...

Reply:
Is this something we care about supporting, or do we declare this "too niche"?

Reply:
We could pass in a wrapper around `&mut Vec<(Readiness, Token)>`, if we wanted. But I think it's not worth adding preemptively.